### PR TITLE
Add NextCloud provider

### PR DESCRIPTION
**Pull request by EMostafaAli**

### CODE REVIEW

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml [hunk 1]
```diff
@@ -633,6 +633,22 @@
     </Environment>
   </Provider>
 
+  <Provider Name="NextCloud" Id="f256847d-c720-4b2f-bda0-69e8f4bceac2"
```
Comment (kevinchalet) [inline src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml]: AFAICT by looking at the official website and the wiki page, the correct casing is `Nextcloud`. Can you please change it?

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml [hunk 2]
```diff
@@ -633,6 +633,22 @@
...
+      <Configuration AuthorizationEndpoint="https://{settings.Tenant}/index.php/apps/oauth2/authorize"
+                   TokenEndpoint="https://{settings.Tenant}/index.php/apps/oauth2/api/v1/token"
+                   UserinfoEndpoint="https://{settings.Tenant}/ocs/v2.php/cloud/user?format=json">
+        <TokenEndpointAuthMethod Value="client_secret_post" />
+        <GrantType Value="authorization_code" />
+        <GrantType Value="refresh_token" />
+      </Configuration>
+    </Environment>
+
+    <Setting PropertyName="Tenant" ParameterName="tenant" Type="String" Required="true"
```
Comment (kevinchalet) [inline src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml]: We generally tend to only use `Tenant` for providers that use a fixed domain but have either a configurable subdomain or virtual path. For providers where the whole host is dynamic, we prefer using `Issuer`. E.g:

https://github.com/openiddict/openiddict-core/blob/a812ca46711d511d1b2dd3b9c237ed85fe4cfbbe/src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml#L26-L27

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml [hunk 3]
```diff
@@ -633,6 +633,22 @@
     </Environment>
   </Provider>
 
+  <Provider Name="NextCloud" Id="f256847d-c720-4b2f-bda0-69e8f4bceac2"
+            Documentation="https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/oauth2.html">
+    <Environment Issuer="https://nextcloud.com/">
+      <Configuration AuthorizationEndpoint="https://{settings.Tenant}/index.php/apps/oauth2/authorize"
+                   TokenEndpoint="https://{settings.Tenant}/index.php/apps/oauth2/api/v1/token"
+                   UserinfoEndpoint="https://{settings.Tenant}/ocs/v2.php/cloud/user?format=json">
+        <TokenEndpointAuthMethod Value="client_secret_post" />
```
Comment (kevinchalet) [inline src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml]: `client_secret_post` is the default token endpoint authentication method for web providers (it's automatically added by `OpenIddict.Client.WebIntegration.Generators`). You can safely remove this line.

**Review by kevinchalet — COMMENTED** [review]
Thanks for your PR! 👍🏻
Do you confirm the provider works fine?

It's worth noting OpenIddict 4.7 will include a built-in claims mapping feature to map well-known claims like the user ID or the username: https://documentation.openiddict.com/guides/contributing-a-new-web-provider.html#if-the-provider-doesnt-support-standard-openid-connect-userinfo-map-the-provider-specific-claims-to-their-claimtypes-equivalent. Since Nextcloud doesn't support OIDC, it's likely they use custom parameters: that would be cool if you could update the built-in handler to support Nextcloud:

https://github.com/openiddict/openiddict-core/blob/a812ca46711d511d1b2dd3b9c237ed85fe4cfbbe/src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.cs#L946-L1153

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml (hunk 1 above)
Comment (EMostafaAli) [reply src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml]: Done

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml (hunk 2 above)
Comment (EMostafaAli) [reply src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml]: Done

**Comment by EMostafaAli** [comment]
> Do you confirm the provider works fine?

Yes, it works.



> Since Nextcloud doesn't support OIDC, it's likely they use custom parameters: that would be cool if you could update the built-in handler to support Nextcloud:

I am not sure if Nextcloud provides claims, this is what I see in the homepage after signing-in

Screenshot from 2023-08-05 01-42-55

**Comment by kevinchalet** [comment]
> I am not sure if Nextcloud provides claims, this is what I see in the homepage after signing-in

If they offer a userinfo endpoint, it's likely at least a user identifier is returned. You don't see it because `AuthenticationController.cs` doesn't preserve these custom claims in the authentication cookie, but you should be able to see them by putting a breakpoint here:

https://github.com/openiddict/openiddict-core/blob/a812ca46711d511d1b2dd3b9c237ed85fe4cfbbe/sandbox/OpenIddict.Sandbox.AspNetCore.Client/Controllers/AuthenticationController.cs#L156

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml [hunk 4]
```diff
@@ -633,6 +633,29 @@
...
+                                  ██ ▀██ ██ ▄▄▄█▄▀█▀▄█▄▄ ▄▄██ ▄▄▀██ █████ ▄▄▄ ██ ██ ██ ▄▄▀██
+                                  ██ █ █ ██ ▄▄▄███ █████ ████ █████ █████ ███ ██ ██ ██ ██ ██
+                                  ██ ██▄ ██ ▀▀▀█▀▄█▄▀███ ████ ▀▀▄██ ▀▀ ██ ▀▀▀ ██▄▀▀▄██ ▀▀ ██
+                                  ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀
+  -->
+
+  <Provider Name="Nextcloud" Id="f256847d-c720-4b2f-bda0-69e8f4bceac2"
+            Documentation="https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/oauth2.html">
+    <Environment Issuer="{settings.Issuer}">
+      <Configuration AuthorizationEndpoint="{CreateAbsoluteUri(settings.Issuer, 'index.php/apps/oauth2/authorize')}"
```
Comment (kevinchalet) [inline src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml:647]: @EMostafaAli FYI, I pushed a commit against your PR branch to make the endpoint URIs computation logic more robust and avoid the double-slash syndrome 😃

**Comment by EMostafaAli** [comment]
Mapping claims has been added.

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.cs [hunk 5]
```diff
@@ -995,6 +995,9 @@ public ValueTask HandleAsync(ProcessAuthenticationContext context)
                 // Mailchimp returns the email address as a custom "login/login_email" node:
                 ProviderTypes.Mailchimp => (string?) context.UserinfoResponse?["login"]?["login_email"],
 
+                // Nextcloud returns the email address as a custom "ocs/data/email" node:
+                ProviderTypes.Nextcloud => (string?) context.UserinfoResponse?["ocs"]?["data"]?["email"],
```
Comment (kevinchalet) [inline src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.cs]: Hum, is the userinfo response wrapped? Could you please share a sample of a userinfo response for Nextcloud? (feel free to redact personal details)

Comment (kevinchalet) [reply src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.cs]: I found the answer to that question in the Nextcloud OpenAPI docs (they do wrap the userinfo details under ocs/data): https://github.com/nextcloud/server/blob/c5f72f0761df8b4e2e8e5c7c3be0bf3a3d17ad55/apps/provisioning_api/openapi.json#L2374-L2388

I pushed a commit that unwraps the userinfo response (which requires changing the claims mapping a bit).

Please give it a final try. Once you confirm it works fine, I'll merge this PR and make sure it's included in the release coming next week.

Thanks!

Comment (kevinchalet) [reply src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.cs]: Note: I documented the unwrapping process here: https://documentation.openiddict.com/guides/contributing-a-new-web-provider.html#unwrap-userinfo-responses-if-necessary.

**Comment by EMostafaAli** [comment]
I confirm it works fine. Thanks.

**Comment by kevinchalet** [comment]
Merged. Congrats for your first contribution to OpenIddict! 🎉

**Pull request merged**